app/controllers/devise/two_factor_authentication_controller.rb

```diff
@@ -12,19 +13,32 @@ class TwoFactorAuthenticationController < DeviseController
     def new
       analytics.track_event('User requested a new OTP code')
 
-      current_user.send_new_otp
+      send_user_otp
+
+      show_direct_otp_prompt
       flash[:notice] = t('devise.two_factor_authentication.user.new_otp_sent')
-      redirect_to user_two_factor_authentication_path(method: 'sms')
     end
 
     def show
       if use_totp?
         show_totp_prompt
       else
-        show_direct_otp_prompt
+        @phone_number = user_decorator.masked_two_factor_phone_number
       end
     end
 
+    def confirm
+      @phone_number = user_decorator.masked_two_factor_phone_number
+      @code_value = current_user.direct_otp if FeatureManagement.prefill_otp_codes?
+    end
+
+    def send_code
+      analytics.track_event("User requested #{current_otp_delivery_method} OTP delivery")
+      send_user_otp
+      show_direct_otp_prompt
+      flash[:success] = t("notices.send_code.#{current_otp_delivery_method}")
+    end
```

Comment:
There was already an analytics event for `User requested a new OTP code` which is a different event than when a user chooses _how_ to receive their OTP. We now have a new event to track.

Just being a different user action alone warrants its own controller method, imo, not to mention that the flash messages and the analytics events are different. 

Yes there is some duplication but I don't think it is unnecessary.